### PR TITLE
[Multi-modal] Add support of Qwen2.5-VL

### PR DESCRIPTION
**Pull request by wwl2755-google**
# Description

This PR adds the initial support for multi-modal input (image-only as the first PR). Together, the support for Qwen2.5-VL with VIT is also included in this PR. 

In this PR, I intentionally separate the mm support apart from the existing logic and skip some plug-in portions (e.g. warm-up) for easier-review. Most part of this PR won't be executed in existing text-only use cases. 

Known issues (will lead to future PRs):

- [x] E2E example/test
- [ ] Unit tests (the util file is now excluded in test coverage and will be included after the unit test PR)
  - [x] Some unit tests are added for tpu_jax_runner but more are coming in the future PR
- [ ] Consolidate some logic with existing ones/ Clean unnecessary parts
- [ ] Procompute warm-up
  - [ ] subtask: check the shape change of t,h,w
- [ ] Larger models (needs TP>1)
- [ ] Reduce HBM consumption (be able to set gpu_memory_utilization to a higher value as default)

CC: @xiangxu-google @kathyyu-google @pengchongjin 

# Tests

Before rebasing on main:
[single prompt](https://paste.googleplex.com/5606753194213376)
[multiple prompts](https://paste.googleplex.com/6724513030733824)

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Comment by wwl2755-google** [comment]
Also cc: @fenghuizhang

File: examples/multi_modal_inference.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+        print("-" * 50)
+    else:
+        yield
+
+
+def parse_args():
+    parser = FlexibleArgumentParser(
+        description="Demo on using vLLM for offline inference with "
+        "vision language models for text generation")
+    parser.add_argument(
```
Comment (xiangxu-google) [inline examples/multi_modal_inference.py:134]: maybe we don't need this

File: examples/multi_modal_inference.py [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+    )
+    parser.add_argument("--num-prompts",
+                        type=int,
+                        default=4,
+                        help="Number of prompts to run.")
+    parser.add_argument(
+        "--modality",
+        type=str,
+        default="image",
+        choices=["image", "video"],
```
Comment (xiangxu-google) [inline examples/multi_modal_inference.py:173]: todo: support video

File: examples/multi_modal_inference.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    )
+    parser.add_argument(
+        "--seed",
+        type=int,
+        default=None,
+        help="Set the seed when initializing `vllm.LLM`.",
+    )
+
+    parser.add_argument(
+        "--image-repeat-prob",
```
Comment (xiangxu-google) [inline examples/multi_modal_inference.py]: remove if not used

File: examples/multi_modal_inference.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+    modality = args.modality
+    mm_input = get_multi_modal_input(args)
+    data = mm_input["data"]
+    questions = mm_input["questions"]
+
+    req_data = model_example_map[model](questions, modality, args)
+
+    # Disable other modalities to save memory
+    default_limits = {"image": 0, "video": 0, "audio": 0}
```
Comment (xiangxu-google) [inline examples/multi_modal_inference.py:225]: why use 0 here

File: examples/multi_modal_inference.py [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+            # Use the same image for all prompts
+            inputs = [{
+                "prompt": prompts[i % len(prompts)],
+                "multi_modal_data": {
+                    modality: data
+                },
+            } for i in range(args.num_prompts)]
+
+    # Add LoRA request if applicable
+    lora_request = (req_data.lora_requests *
```
Comment (xiangxu-google) [inline examples/multi_modal_inference.py]: pls remove as we don't support lora

File: tpu_commons/models/jax/attention_metadata.py [hunk 6]
```diff
@@ -33,3 +34,5 @@ class AttentionMetadata(object):
     num_seqs: jax.Array = None
     # (1,)
     num_slices: jax.Array = None
+    # (padded_total_num_scheduled_tokens,)
+    input_mrope_positions: jax.Array = None
```
Comment (xiangxu-google) [inline tpu_commons/models/jax/attention_metadata.py]: seems it's not used, pls remove

File: tpu_commons/models/jax/model_loader.py [hunk 7]
```diff
@@ -189,9 +193,29 @@ def run_compute_logits(graphdef, state, *args):
         model = nnx.merge(graphdef, state)
         return model.compute_logits(*args)
 
+    # Multi-modal support
```
Comment (xiangxu-google) [inline tpu_commons/models/jax/model_loader.py]: add more comments for these two functions, they are both only for multi-modal models

File: tpu_commons/models/jax/utils/multi_modal_utils.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        return _merge_multimodal_embeddings(
+            inputs_embeds,
+            jnp.isin(input_ids, placeholder_token_id),
+            multimodal_embeddings,)
+
+    return _merge_multimodal_embeddings(
+        inputs_embeds,
+        (input_ids == placeholder_token_id),
+        multimodal_embeddings,
+    )
```
Comment (xiangxu-google) [inline tpu_commons/models/jax/utils/multi_modal_utils.py]: pls confirm running pre-commit to reformat code

File: tpu_commons/models/jax/qwen2_5_vl.py [hunk 9]
```diff
@@ -0,0 +1,918 @@
...
+
+
+class Qwen2_5_VisionMLP(nnx.Module):
+
+    def __init__(self, config: Qwen2_5_VLVisionConfig, dtype: jnp.dtype,
+                 rngs: nnx.Rngs):
+        in_features = config.hidden_size
+        hidden_features = config.intermediate_size
+        act_fn = modeling_flax_utils.ACT2FN[config.hidden_act]
+        self.gate_proj = nnx.Linear(
```
Comment (xiangxu-google) [inline tpu_commons/models/jax/qwen2_5_vl.py:97]: pls add model sharding support

File: tpu_commons/models/jax/qwen2_5_vl.py [hunk 10]
```diff
@@ -0,0 +1,918 @@
...
+
+        segment_ids = None
+        if use_fullattn:
+            segment_ids_val = (jnp.arange(padded_T)
+                           >= T_attn).astype(jnp.int32).reshape(B, -1)
+            segment_ids = SegmentIds(q=segment_ids_val, kv=segment_ids_val)
+        else:
+            segment_ids = generate_window_segment_ids(cu_window_seqlens, T_attn, padded_T)
+
+        output = flash_attention(q,
```
Comment (xiangxu-google) [inline tpu_commons/models/jax/qwen2_5_vl.py]: pls use sharded version

File: tpu_commons/models/jax/qwen2_5_vl.py [hunk 11]
```diff
@@ -0,0 +1,918 @@
...
+        cu_seqlens_tmp = cu_seqlens_tmp.astype(jnp.int32)
+
+        # NOTE (wenlong): Pytorch code uses this to reduce replication,
+        # but I don't think there is a need here, plus it would cause problem in JIT
+        # Please refer here if there is a problem down-stream
+        # cu_seqlens_tmp = jnp.unique(cu_seqlens_tmp)
+
+        return index_new, cu_seqlens_tmp
+
+    def get_rope_by_thw(self, t, h, w):
```
Comment (xiangxu-google) [inline tpu_commons/models/jax/qwen2_5_vl.py:560]: pls check the shape change of t,h,w

File: tpu_commons/models/jax/qwen2_5_vl.py [hunk 12]
```diff
@@ -0,0 +1,918 @@
...
+        return self.language_model.compute_logits(hidden_states)
+
+    def load_weights(self, rng_key: jax.Array) -> None:
+        self.rng = nnx.Rngs(rng_key)
+        self.language_model.rng = self.rng
+
+        # Key: path to a HF layer weight
+        # Value: a tuple of (path to a nnx layer weight, nnx weight sharding)
+        mappings = {
+            "model.embed_tokens": ("language_model.embed.embedding", ("model",
```
Comment (xiangxu-google) [inline tpu_commons/models/jax/qwen2_5_vl.py]: pls move sharding annonation inside the model

File: tpu_commons/platforms/tpu_jax.py [hunk 13]
```diff
@@ -136,14 +136,16 @@ def check_and_update_config(cls, vllm_config: VllmConfig) -> None:
...
         # NOTE(xiang): convert dtype to jnp.dtype
-        if not isinstance(vllm_config.model_config.dtype, str):
-            logger.warning(
-                "The model dtype is not properly set for JAX backend. "
-                "Overwriting it to bfloat16")
-            vllm_config.model_config.dtype = jnp.bfloat16
-        else:
-            vllm_config.model_config.dtype = _DTYPE.get(
-                vllm_config.model_config.dtype, jnp.bfloat16)
+        # NOTE(wenlong): skip this logic for mm model preprocessing
```
Comment (xiangxu-google) [inline tpu_commons/platforms/tpu_jax.py:142]: add more explanation why need to use torch types for mm models

File: tpu_commons/runner/jax/tpu_jax_runner.py [hunk 14]
```diff
@@ -226,7 +265,6 @@ def get_supported_tasks(self) -> tuple[SupportedTask, ...]:
 
     def get_kv_cache_spec(self):
         # TODO(xiang): this hack tricks engine core to init successfully
-        import torch
```
Comment (xiangxu-google) [inline tpu_commons/runner/jax/tpu_jax_runner.py:217]: pls revert this

File: tpu_commons/runner/jax/tpu_jax_runner.py [hunk 15]
```diff
@@ -668,15 +904,40 @@ def _execute_model(
...
+
+        # multi-modal support
+        if self.is_multimodal_model:
+            # Run the multimodal encoder if any.
+            self._execute_mm_encoder(scheduler_output)
+            mm_embeds = self._gather_mm_embeddings(scheduler_output)
+        else:
+            mm_embeds = []
+
+        input_ids, inputs_embeds = self._get_model_inputs(input_ids, mm_embeds)
```
Comment (xiangxu-google) [inline tpu_commons/runner/jax/tpu_jax_runner.py:911]: pls add some comments to explain the diff between mm and text-only models

File: tpu_commons/runner/jax/tpu_jax_runner.py [hunk 16]
```diff
@@ -641,6 +685,197 @@ def insert_request_with_kv_cache(
         self.requests[req_id] = req_state
         self.input_batch.add_request(req_state)
 
+    def _calc_mrope_positions(self, scheduler_output: "VllmSchedulerOutput"):
```
Comment (xiangxu-google) [inline tpu_commons/runner/jax/tpu_jax_runner.py:675]: pls check if where is the mrope position is calculated, if it's calculated inside the model, pls remove this

File: examples/multi_modal_inference.py (hunk 4 above)
Comment (wwl2755-google) [reply examples/multi_modal_inference.py:225]: It initialized all modalities to be 0s and merged with the specified one later (i.e. inside run_qwen2_5_vl(), it will make "image" to be 1).

File: examples/multi_modal_inference.py (hunk 1 above)
Comment (wwl2755-google) [reply examples/multi_modal_inference.py:134]: Deleted.

File: examples/multi_modal_inference.py (hunk 2 above)
Comment (wwl2755-google) [reply examples/multi_modal_inference.py:173]: Updated. Added some comments

File: examples/multi_modal_inference.py (hunk 3 above)
Comment (wwl2755-google) [reply examples/multi_modal_inference.py]: Deleted

File: examples/multi_modal_inference.py (hunk 5 above)
Comment (wwl2755-google) [reply examples/multi_modal_inference.py]: Removed.

File: tpu_commons/models/jax/model_loader.py (hunk 7 above)
Comment (wwl2755-google) [reply tpu_commons/models/jax/model_loader.py]: More comments added.

File: tpu_commons/platforms/tpu_jax.py (hunk 13 above)
Comment (wwl2755-google) [reply tpu_commons/platforms/tpu_jax.py:142]: More explanation added.

File: tpu_commons/runner/jax/tpu_jax_runner.py (hunk 14 above)
Comment (wwl2755-google) [reply tpu_commons/runner/jax/tpu_jax_runner.py:217]: Updated.

File: tpu_commons/runner/jax/tpu_jax_runner.py (hunk 16 above)
Comment (wwl2755-google) [reply tpu_commons/runner/jax/tpu_jax_runner.py:675]: I just realized that it should be actually used. The rope we inside the qwen2_5_vl.py is calculating the positional embedding inside the VIT. In the forward pass after merging image embeddings and text embeddings, we still need the positional index to capture the image information. 

Details can be seen at: https://github.com/vllm-project/vllm/blob/main/vllm/v1/worker/gpu_model_runner.py#L1541-L1544, https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/qwen2_5_vl.py#L1095-L1099 

That part is missing for now. So the code still needs some update.

File: tpu_commons/models/jax/qwen2_5_vl.py (hunk 12 above)
Comment (wwl2755-google) [reply tpu_commons/models/jax/qwen2_5_vl.py]: Updated.

File: tpu_commons/models/jax/qwen2_5_vl.py (hunk 9 above)
Comment (wwl2755-google) [reply tpu_commons/models/jax/qwen2_5_vl.py:97]: Updated.

File: tpu_commons/models/jax/attention_metadata.py (hunk 6 above)
Comment (wwl2755-google) [reply tpu_commons/models/jax/attention_metadata.py]: It is actually used in the forward pass. Updated in the latest commit.

File: tpu_commons/models/jax/qwen2_5_vl.py (hunk 11 above)
Comment (wwl2755-google) [reply tpu_commons/models/jax/qwen2_5_vl.py:560]: Updated in the Todo in this PR. Plan to send a PR later together with the precompile run

File: tpu_commons/runner/jax/tpu_jax_runner.py (hunk 15 above)
Comment (wwl2755-google) [reply tpu_commons/runner/jax/tpu_jax_runner.py:911]: Updated.

File: tpu_commons/models/jax/qwen2_5_vl.py (hunk 10 above)
Comment (wwl2755-google) [reply tpu_commons/models/jax/qwen2_5_vl.py]: Updated.

File: tpu_commons/models/jax/utils/multi_modal_utils.py (hunk 8 above)
Comment (wwl2755-google) [reply tpu_commons/models/jax/utils/multi_modal_utils.py]: Updated. Local test pre-commit passed after rebasing.

**Comment by wwl2755-google** [comment]
It works fine with the latest main at: 0c67f6158e8001e425e30beaeb5fbf8f3d7a92b7

https://paste.googleplex.com/5288634630275072

File: tpu_commons/runner/jax/tpu_jax_runner.py (hunk 16 above)
Comment (wwl2755-google) [reply tpu_commons/runner/jax/tpu_jax_runner.py:675]: The code is updated and mrope calculation is verified.

**Comment by xiangxu-google** [comment]
https://buildkite.com/tpu-commons/tpu-commons-ci/builds/1475

**Comment by wwl2755-google** [comment]
Qwen2.5-VL (multi-modal), llama3.1-8B and qwen2.5-1.5B works fine in offline serving. https://paste.googleplex.com/5899696270475264

Tested locally after rebasing on: 61d5d4b0c32922cfc56c9c553bff4ca98e4bec97

**Comment by wwl2755-google** [comment]
After fixing precompile backbone commit: (should solve the issue of re-compilation in CI) cc: @xiangxu-google 

https://paste.googleplex.com/5723874234466304

**Comment by wwl2755-google** [comment]
test_model_loader fixed in the latest commit.

Verified locally: https://paste.googleplex.com/5659959349739520

**Comment by wwl2755-google** [comment]
Locally tested unit tests: https://paste.googleplex.com/4992754199035904

**Comment by wwl2755-google** [comment]
All tests mentioned above passed locally after rebasing on : 2e734f32323d7324445a506947cdfc2b9f212f4b

**Comment by wwl2755-google** [comment]
Jax+vLLM is fixed by latest commit.

Verified: https://paste.googleplex.com/6357418350739456

**Comment by wwl2755-google** [comment]
fixed unit test for get_vllm_model. Verified: https://paste.googleplex.com/6302452433027072
